src/VisualStudio/Core/Def/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs

```diff
@@ -224,6 +231,70 @@ public int RemoveAdornments()
 
                 return VSConstants.S_OK;
             }
+
+            // GetOutline is called every time a new code window is created. Whenever we switch to a different window, it is guaranteed
+            // that ReleaseOutline will be called on the old window before GetOutline is called for the new window. 
+            int IVsDocOutlineProvider.GetOutline(out IntPtr phwnd, out IOleCommandTarget? ppCmdTarget)
+            {
+                var languageServiceBroker = _languageService.Package.ComponentModel.GetService<ILanguageServiceBrokerShim>();
+                var threadingContext = _languageService.Package.ComponentModel.GetService<IThreadingContext>();
+                var asyncListenerProvider = _languageService.Package.ComponentModel.GetService<IAsynchronousOperationListenerProvider>();
+                var asyncListener = asyncListenerProvider.GetListener(FeatureAttribute.DocumentOutline);
+                var editorAdaptersFactoryService = _languageService.Package.ComponentModel.GetService<IVsEditorAdaptersFactoryService>();
+
+                threadingContext.ThrowIfNotOnUIThread();
+
+                // Assert that the previous Document Outline Control and host have been freed. 
+                Contract.ThrowIfFalse(_documentOutlineControl is null);
+                Contract.ThrowIfFalse(_documentOutlineViewHost is null);
+
+                _documentOutlineControl = new DocumentOutlineControl(
+                        languageServiceBroker, threadingContext, asyncListener, editorAdaptersFactoryService, _codeWindow);
+
+                _documentOutlineViewHost = new ElementHost
+                {
+                    Dock = DockStyle.Fill,
+                    Child = _documentOutlineControl
+                };
+
+                phwnd = _documentOutlineViewHost.Handle;
+                ppCmdTarget = null;
+
+                Logger.Log(FunctionId.DocumentOutline_WindowOpen);
+
+                return VSConstants.S_OK;
+            }
+
+            int IVsDocOutlineProvider.ReleaseOutline(IntPtr hwnd, IOleCommandTarget pCmdTarget)
+            {
+                var threadingContext = _languageService.Package.ComponentModel.GetService<IThreadingContext>();
+                threadingContext.ThrowIfNotOnUIThread();
+
+                // Assert that we are not attempting to double free the Document Outline Control and host.
+                Contract.ThrowIfNull(_documentOutlineViewHost);
+                Contract.ThrowIfNull(_documentOutlineControl);
+
+                _documentOutlineViewHost.SuspendLayout();
+                _documentOutlineControl.Dispose();
+                _documentOutlineControl = null;
+                _documentOutlineViewHost.Child = null;
+                _documentOutlineViewHost.Parent = null;
+                _documentOutlineViewHost.Dispose();
+                _documentOutlineViewHost = null;
```

Comment:
Is it necessary to be interleaving the cleanup of _documentOutlineViewHost and _documentOutlineControl in this way? If so, it'd be worth adding some comments as to why that's critical.

Reply:
when `ElementHost` is disposed it attempts to remove itself from its parent container. This operation will fail unless we do additional operations as the shell is already in the process of removing us when we are called. 

For the Dispose call to succeed we need to dispose and null out all of the `ElementHost`'s resources before calling dispose. This causes `ElementHost` to no-op and just dispose of its own native resources. This is apparently a common gotcha with disposal of winforms/wpf elements interleaved with native components.